src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java

```diff
@@ -1176,26 +992,73 @@ public boolean checkOrRequestItems(final boolean useItemDamage, @Nullable final
             return false;
         }
         boolean allClear = true;
-        for (final @Nullable ItemStack stack : items)
+        for (final @Nullable ItemStack tempStack : items)
         {
-            if (stack == null || stack.getItem() == null)
+            if (ItemStackUtils.isEmpty(tempStack))
             {
                 continue;
             }
-            final int countOfItem;
-            if (useItemDamage)
+            final ItemStack stack = tempStack.copy();
+
+
+            final int itemDamage = useItemDamage ? stack.getItemDamage() : -1;
+            final int countOfItem = worker.getItemCountInInventory(stack.getItem(), itemDamage);
+
+            if (countOfItem < 1)
             {
-                countOfItem = worker.getItemCountInInventory(stack.getItem(), stack.getItemDamage());
+                final int itemsLeft = ItemStackUtils.getSize(stack) - countOfItem;
+                @NotNull final ItemStack requiredStack = new ItemStack(stack.getItem(), itemsLeft, itemDamage);
+                getOwnBuilding().addNeededItems(requiredStack);
+                allClear = false;
             }
             else
             {
-                countOfItem = worker.getItemCountInInventory(stack.getItem(), -1);
+                getOwnBuilding().clearNeededItems();
             }
-            if (countOfItem < 1)
+        }
+        if (allClear)
+        {
+            return false;
+        }
+        itemsNeeded.clear();
+        Collections.addAll(itemsNeeded, items);
+        return true;
+    }
+
+    /**
+     * Require that items are in the workers inventory.
+     * This safeguard ensures you have said items before you execute a task.
+     * This doesn't stop execution on false returned.
+     * It does request asynchronously and make sure nothing is added twice.
+     *
+     * @param useItemDamage compare the itemDamage of the values.
+     * @param items         the items needed
+     * @return false if they are in inventory
+     */
+    public boolean checkOrRequestItemsAsynch(final boolean useItemDamage, @Nullable final ItemStack... items)
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Provide multiple methods instead of using "useItemDamage" to determine which action to take. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2301)